w3af/plugins/grep/vulners_scanner.py

```diff
@@ -0,0 +1,219 @@
+        # Get fresh rules from Vulners.
+        try:
+            self.rules_table = self.get_vulners_api().rules()
+            # Adapt it for MultiRe structure [(regex,alias)] removing regex duplicated
+            regexAliases = collections.defaultdict(list)
+            for softwareName in self.rules_table:
+                regexAliases[self.rules_table[softwareName].get('regex')] += [softwareName]
+            # Now create fast RE filter
+            # Using re.IGNORECASE because w3af is modifying headers when making RAW dump. Why so? Raw must be raw!
+            self._multi_re = MultiRE(((regex, regexAliases.get(regex)) for regex in regexAliases), re.IGNORECASE)
+        except Exception as e:
+            self.rules_table = None
+            error_message = 'Vulners plugin failed to init with error: %s'
+            om.out.error(error_message % e)
+
+    def get_vulners_api(self):
+        # Lazy import. Just not to make it in the __init__
+        # Other way API will try to communicate Vulners right at the w3af startup
+        if not self._vulners_api:
+            # Don't forget to setup Vulners API key if you have one
+            self._vulners_api = vulners.Vulners(api_key= self._vulners_api_key)
+        return self._vulners_api
+
+    def check_vulners(self, software_name, software_version, check_type):
+        vulnerabilities = {}
+        cached_result = self._vulnerability_cache.get((software_name, software_version, check_type))
+        if cached_result:
+            return cached_result
+        # Ask Vulners about vulnerabilities
+        if check_type == 'software':
+            vulnerabilities = self.get_vulners_api().softwareVulnerabilities(software_name, software_version)
+        elif check_type == 'cpe':
+            cpe_string = "%s:%s" % (software_name, software_version)
+            vulnerabilities = self.get_vulners_api().cpeVulnerabilities(cpe_string.encode())
+        self._vulnerability_cache[(software_name, software_version, check_type)] = vulnerabilities
+        return vulnerabilities
+
+    def get_severity(self, cvss_score):
+        # This table can be actually modified
+        # That's only my opinions on the severity vs CVSS.score
+        # Maybe move it to the options later?
+        cvss_score = cvss_score * 10
+        if cvss_score in range(20, 30):
+            return severity.LOW
+        elif cvss_score in range(30, 70):
+            return severity.MEDIUM
+        elif cvss_score in range(70, 100):
+            return severity.HIGH
+        else:
+            return severity.INFORMATION
+
+    def grep(self, request, response):
+        """
+        Plugin entry point, search for vulnerable software banners  in web application HTTP responses.
+        :param request: The HTTP request object.
+        :param response: The HTTP response object
+        :return: None
+
+        """
+        # Lazy update rules if it's first start of the plugin
+        if not self.rules_updated:
```

Comment:
There is a race condition here, where multiple threads will be running the the rules will be updated 2 or 3 times.

Use (check in other plugins, not 100% sure about the name):

```python
with self._plugin_lock:
    if not self.rules_updated:
        ...
```

Reply:
Applied.